src/launchpad/service.py

```diff
@@ -78,58 +127,396 @@ def handle_kafka_message(self, payload: PreprodArtifactEvents) -> None:
+        organization_id: str,
+        error_code: int,
+        error_message: ProcessingErrorMessage,
+        detailed_error: str | None = None,
+    ) -> None:
+        """Update artifact with error information."""
+        try:
+            logger.info(f"Updating artifact {artifact_id} with error code {error_code}")
+
+            # Use detailed error message if provided, otherwise use enum value
+            final_error_message = f"{error_message.value}: {detailed_error}" if detailed_error else error_message.value
+
+            # Log error to datadog with tags for better monitoring
+            if self._statsd:
+                self._statsd.increment(
+                    "launchpad.artifact.processing.error",
+                    tags=[
+                        f"error_code:{error_code}",
+                        f"error_type:{error_message.name}",
+                        f"project_id:{project_id}",
+                        f"organization_id:{organization_id}",
+                    ],
+                )
+
+            result = sentry_client.update_artifact(
+                org=organization_id,
+                project=project_id,
+                artifact_id=artifact_id,
+                data={"error_code": error_code, "error_message": final_error_message},
+            )
+
+            if "error" in result:
+                logger.error(f"Failed to update artifact with error: {result['error']}")
+            else:
+                logger.info(f"Successfully updated artifact {artifact_id} with error information")
+
+        except Exception as e:
+            logger.error(f"Failed to update artifact {artifact_id} with error information: {e}", exc_info=True)
+
+    def _handle_download_error(
+        self, sentry_client: SentryClient, artifact_id: str, project_id: str, organization_id: str, error_msg: str
+    ) -> None:
+        """Handle download error by logging and updating artifact."""
+        logger.error(error_msg)
+        # Extract just the error details for the database, the enum prefix will be added automatically
+        detailed_error = error_msg.split(": ", 1)[1] if ": " in error_msg else error_msg
+        self._update_artifact_error(
+            sentry_client,
+            artifact_id,
+            project_id,
+            organization_id,
+            ERROR_CODE_ARTIFACT_PROCESSING_ERROR,
+            ProcessingErrorMessage.DOWNLOAD_FAILED,
+            detailed_error,
+        )
+        raise RuntimeError(error_msg)
+
+    def _download_artifact(
+        self,
+        sentry_client: SentryClient,
+        artifact_id: str,
+        project_id: str,
+        organization_id: str,
+    ) -> tuple[bytes, int]:
+        """Download artifact from Sentry and validate response."""
+        logger.info(f"Downloading artifact {artifact_id}...")
 
         try:
-            # Wait for shutdown signal
-            await self._shutdown_event.wait()
+            download_result = sentry_client.download_artifact(
```

Comment:
Here it might be better to do:

```python
with tempfile.TemporyFile() as tf:
  sentry_client.download_artifact(org=organization_id, project=project_id, artifact_id=artifact_id, out=tf)
  artifact = ArtifactFactory.fromFile(tf)
  ...
  
```